accepted/2022/FunctionPointers.md

```diff
@@ -628,12 +636,41 @@ namespace System.Reflection
 }
 ```
 
+## Support Type.GetMethod()
+If a user tries to retrieve a Method through [`Type.GetMethod()`](https://learn.microsoft.com/dotnet/api/system.type.getmethod) that has function pointer arguments specified by the `Type[]` passed into that method, the current design throws `NotSupportedException` since we don't have a good way to specify the custom modifiers for each parameter in the API (see below -- requires "MethodSignature").
```

Comment:
FWIW, C# doesn't actually allow things to overload if the only difference is in the modifiers that shouldn't matter.

```cs
unsafe class Foo
{
    // Fine
    void A(delegate* unmanaged[Stdcall]<int, void> x) => throw null;
    void A(delegate* unmanaged[Cdecl]<int, void> x) => throw null;

    // Doesn't compile
    void B(delegate*<ref int, void> x) => throw null;
    void B(delegate*<out int, void> x) => throw null;
}
```

So if we did care about the callconv but nothing else, it would be okay for C#. It would still not be okay for other languages, but that's how it is with modifiers right now in general anyway - we ignore them in many positions.

The more I think about it, the more it feels like the `out`/`ref` bits on function pointers should have been encoded as a custom attribute, not as a modifier - same way C# encodes nullability information. The custom modifier throws off a lot of things. If we could go back on that, it would probably help.

Reply:
> The more I think about it, the more it feels like the out/ref bits on function pointers should have been encoded as a custom attribute, not as a modifier - same way C# encodes nullability information. The custom modifier throws off a lot of things. If we could go back on that, it would probably help.

Since function pointers can't have attributes or pseudo-attributes like `[in]` I suppose it was the most straightforward to use modopts. However, I wonder what it would look like if we could go back on ECMA-335 instead and change the somewhat rigid spec of how attributes, pseudo-attributes and modifiers each can\can't be applied to the various typedecls\refs\signatures\parameters.

Reply:
> Since function pointers can't have attributes or pseudo-attributes like `[in]` I suppose it was the most straightforward to use modopts.

We didn't use modopts to encode nullability:

```csharp
class Program
{
    static void Method1(List<object?> param) { }
    static void Method2(List<object> param) { }
}
```

This has the same problem - how C# can keep of the track of the question mark in `Method1`? We could put a modopt on the `object` part, but that affects overload resolution. Instead this is encoded in a custom attribute on the method. An encoding was invented to have this information in a sidecar, no matter how nested the question mark needs to be.

The `ref`/`out`/`in`/`readonly` etc. bits should have gone into a similar sidecar so that they don't mess with overload resolution.